### PR TITLE
Access local engine with an public url (with ngrok)

### PR DESCRIPTION
**Pull request by oeway**
This PR is to support accessing the engine publicly through ngrok proxy.

### CODE REVIEW

File: imjoy/services/ngrok.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import sys
+import os
+import platform
+import urllib.request
+import zipfile
+import stat
+import subprocess
+import json
+
+def detectHostArchitecture():
```
Comment (MartinHjelmare) [inline imjoy/services/ngrok.py]: Please use snakecase names.

File: imjoy/services/ngrok.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        'linux-amd64':        cdn + cdnPath + 'linux-amd64.zip',
+        'win32-386':    cdn + cdnPath + 'windows-386.zip',
+        'win32-amd64':        cdn + cdnPath + 'windows-amd64.zip',
+        'freebsd-386':    cdn + cdnPath + 'freebsd-386.zip',
+        'freebsd-amd64':    cdn + cdnPath + 'freebsd-amd64.zip'
+    }
+
+    url = cdnFiles.get(arch)
+    if url is None:
+        raise Exception('ngrok - platform ' + arch + ' is not supported.')
```
Comment (MartinHjelmare) [inline imjoy/services/ngrok.py]: It's better to raise app specific exceptions. Ie create a base exception class that inherits from `Exception` that we can subclass into different exceptions as needed.

That will make it easier to differentiate exceptions.

File: imjoy/services/ngrok.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    return url
+
+
+def download_ngrok(targetdir):
+    # check if data has been downloaded already
+    zipPath = os.path.join(targetdir, 'ngrok.zip')
```
Comment (MartinHjelmare) [inline imjoy/services/ngrok.py]: Can we use pathlib here?

**Comment by oeway** [comment]
> Would we benefit from a python wrapper for ngrok?
> 
> https://github.com/alexdlaird/pyngrok

Totally missed that, it seems I am reinventing the wheel. I tried to find those but missed that one.

OK, I will just switch to `pyngrok` if it works.

File: imjoy/engine.py [hunk 4]
```diff
@@ -50,8 +51,21 @@ def setup(self):
...
 
+    def start_proxy(self):
+        """Start the reverse proxy for public access."""
+        try:
+            from pyngrok import ngrok
+
+            public_url = ngrok.connect(port=self.opt.port)
+            public_url = public_url.replace("http", "https")
+            self.public_url = public_url
+        except:
```
Comment (MartinHjelmare) [inline imjoy/engine.py]: We should except the pyngrok error here:
https://pyngrok.readthedocs.io/en/latest/api.html#pyngrok.exception.PyngrokError

Comment (oeway) [reply imjoy/engine.py]: Yes, I can add that one. However, I am not confident enough to make sure that's the only error type can happen. We cannot afford that crashing the engine because the proxy failed to start. For example `import pyngrok` may also fail.

File: imjoy/runners/subprocess.py [hunk 5]
```diff
@@ -744,6 +744,13 @@ def process_finish(pid=None, cmd=None):
         stop_callback(False, "Plugin process failed to start")
         return False
     # env = os.environ.copy()
+
+    # Replace python to python3
```
Comment (MartinHjelmare) [inline imjoy/runners/subprocess.py]: Is this change required for ngrok? Otherwise I suggest we move it to a different PR.

File: imjoy/engine.py (hunk 4 above)
Comment (MartinHjelmare) [reply imjoy/engine.py]: Then make two except cases.

File: imjoy/runners/subprocess.py (hunk 5 above)
Comment (oeway) [reply imjoy/runners/subprocess.py]: You are right, it's not related.  I was testing a failed plugin.

**Comment by oeway** [comment]
There is another issue with using ngrok, by default, the connection can fail if a certain time limit exceeded (https://ngrok.com/upgrade).

We may need to host our own server, however, we need a CA certificate to enable https connection.

Not sure if there is easier way.

File: imjoy/engine.py (hunk 4 above)
Comment (oeway) [reply imjoy/engine.py]: Would you do something like this?
```python
        try:
            from pyngrok import ngrok
            from pyngrok.exception import PyngrokError
            try:
                public_url = ngrok.connect(port=self.opt.port)
                public_url = public_url.replace("http", "https")
                self.public_url = public_url
            except PyngrokError:
                self.logger.error("ngrok failed to start.")
        except ImportError:
            self.logger.error("Failed to import pyngrok")
```

Comment (oeway) [reply imjoy/engine.py]: Well, perhaps in this case,  I can just remove the try block to let it crash, the user can turn of the feature with the `--public` anyway.

Comment (MartinHjelmare) [reply imjoy/engine.py]: When would there be an import error?

I would return early if we can't import pyngrok.

But it would be better if we knew earlier that pyngrok is installed.

**Comment by oeway** [comment]
Well, ngrok seems not good enough for us, it requires login and has limitations on time and connection number which are not easy to get rid of. The other thing is that the url are changing every time you start it.

I am now looking into an alternative solution https://telebit.cloud/ which seems very easy to use.
